Feat/pokemon list

Matthieu-EY opened the pull request:
Dans cette PR:
- La page d'accueil a été implémentée, et consiste maintenant en une grille infinie contenant chaque Pokémon. En cliquant dessus, l'utilisateur est redirigé vers la page Détail Pokémon.
- La page détail Pokémon a été largement revu, de manière à afficher plus de détails sur le Pokémon.

Page d'accueil/Liste Pokémon:
<img width="1001" height="556" alt="Capture d’écran 2025-07-15 à 14 31 14" src="https://github.com/user-attachments/assets/62709342-46e6-47d4-bc1a-ca99ddcd7815" />

Page de détail Pokémon:
<img width="1113" height="763" alt="Capture d’écran 2025-07-11 à 16 52 23" src="https://github.com/user-attachments/assets/8df41e94-cb5e-4cb9-ba72-c232506c95a6" />

Ce qui reste à faire:
- Animation lors du survol d'un Pokémon dans la grille ?
- Améliorer la présentation des commandes (homme/femme, shiny/non shiny) sur la page de détail Pokémon
- Revoir le design de la grille et de chaque Pokémon dans cette grille
- Implémenter l'API maison afin de pouvoir stocker des Pokémon dans son équipe
- Implémenter un système de combat ?


Dépendante de la PR [Page détail Pokémon](https://github.com/Matthieu-EY/nextjs-pokemon/pull/1)
Cette PR ne sera donc merge qu'une fois que l'autre sera merge et review en bonne et due forme

Matthieu-EY committed:
feat: page d'accueil, liste de pokemons

Matthieu-EY committed:
feat: page liste pokemon, affichage en grid

Matthieu-EY committed:
fix: pokemon name overflow fix

Matthieu-EY committed:
feat: pokemon type style

- Les types de pokémon sont maintenant affichés avec un joli design.
- Les images des types serons aussi utiles par la suite.

Matthieu-EY committed:
chore: linting prettier

Matthieu-EY committed:
Merge branch 'api-poke' into feat/pokemon-list

Matthieu-EY committed:
fix: retours PR sur nouvelle branche

Matthieu-EY committed:
chore: icon updated

Matthieu-EY committed:
fix: pokemon detail component rename

Matthieu-EY committed:
feat: buttons for male/female & shiny

Matthieu-EY committed:
feat: pokemondetail, icons male/female added

Matthieu-EY committed:
feat pokemon detail, info on pokemon added

The PokemonDetail page now shows:
- The height & weight of the Pokemon
- The cries with a possibility to play the sounds
- The abilities the Pokemon has.

Matthieu-EY committed:
feat: pokemon list, infinite scroll

The pokemon list page now features infinite scroll, as the request is made when the user scrolls to the bottom, giving the impression scroll is infinite.

Matthieu-EY committed:
fix: infinite scroll always triggered on render fixed

Matthieu-EY committed:
fix: slight changes

Matthieu-EY committed:
Squashed commit of the following:

commit 8c66b6fbb71de945af85077f006a6606f9185530
Merge: 36bcd56 2c2c8aa
Author: Ruben De Andrade <28705409+RubenDeAndrade@users.noreply.github.com>
Date:   Tue Jul 15 11:18:37 2025 +0200

    Merge pull request #1 from Matthieu-EY/api-poke

    Page détail Pokémon

commit 2c2c8aaadeb4d46188816a62dbf53f2b643f7e08
Author: Matthieu-EY <matthieu.sirvan@fabernovel.ey.com>
Date:   Fri Jul 11 10:40:46 2025 +0200

    fix: typescript param error fixed

    - Typescript error on param 'await' fixed.

Matthieu-EY committed:
feat: search by name or by type

Matthieu-EY commented:
Prêt à review.

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,138 +1,169 @@
 'use client';
```
Ce composant devrait être un server component, dans lequel tu devrais fetch une premiere fois une liste de pokemon, afin d'améliorer le First Contentful Paint.
Puis trouver la solution qui permettrait d'hydrater useInfiniteQuery côté client.
Tu devras possiblement séparer la partie serveur de la partie client.
A l'avenir, tu devras réfléchir côté serveur dans un premier temps, avant de penser côté client, que ce soit dans le fetching de données, l'authent, l'accès à des données sensibles, les server actions, ou bien le rendu de composants.

Matthieu-EY replied:
C'est fait.
Maintenant, la page Liste Pokémon (accueil) est un Server component, comme la page de détail.
Les éléments interactifs sont des Client component isolés, utilisés par les pages et les Server component.
Cette hiérarchie permet de séparer le travail entre client et server and améliorer le temps de chargement des sites.
Cela permet aussi d'utiliser le SSR offert par Next.JS à fond.

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,138 +1,169 @@
+import { getTokenFromUrl } from '~/utils/get-token-from-url';
+import { useCallback, useRef, useState } from 'react';
+
+const types = [
```
Il faudrait sûrement typé ce tableau. C'est des données utilisés dans PokéAPI ?

Matthieu-EY replied:
Il s'agit de tous les types de Pokémon possible.
Cela m'est utile pour afficher les types dans le select "Search by type".
Je pensais à exporter ces données dans un fichier JSON à part.

Matthieu-EY replied:
Note to self:
Si PokeAPI vient à être modifié, ou un autre type est ajouté par Nintendo, alors le site possèdera encore les anciens types.
D'un autre côté, faire un appel constant pour toutes les connexions paraît déraisonnable.
Il s'agit donc de faire un compromis entre praticité et s'assurer que le site ne devienne pas "vieux", déconnecté des mises à jour.
MAIS, les types de Pokémons n'ont pas été modifiés depuis des lustres, donc cela vaut t-il vraiment la peine ?

Matthieu-EY replied:
Pire encore, est le cas des types "???" et "stellaire".
Le type "???" est un type apparu dans les générations <4 pour catégoriser des Pokémons cachés. Ce type est introuvable et inutilisable.
Le type "stellaire" est un type apparu dans la génération 9 (quasiment la dernière) qui correspond à une transformation temporaire du Pokémon.
Ces 2 types sont évidemment problématiques à implémenter, voire ne doivent pas être utilisés/répertorié. Aucun Pokémon n'a le type Stellaire par exemple. Donc inutile de l'ajouter au select.
A la vue de ces cas, je recommande fortement de conserver la liste en local de nom de types.
Simple (KISS), ne scale pas avec les nouvelles générations, mais au vu des nouveaux types ajoutés par les dernières générations, c'est peut-être pour le mieux.

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,138 +1,169 @@
+
+  // name inside "search by name" input
+  const [searchedName, setSearchedName] = useState('');
+  const [searchedType, setSearchedType] = useState('All');
```
searchedType: string 
Il faudrait améliorer le typage également. Ici je peux modifier avec n'importe quelle string, néanmoins ce ne doit pas être la volonté de cet état.

Matthieu-EY replied:
Implémenté.

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,138 +1,169 @@
+  const pokemonsQuery = trpc.poke.list.useInfiniteQuery(
     {
-      limit: 5,
+      limit: 50,
```
créer une constante avec un nom explicite pour les valeurs fixes.

Matthieu-EY replied:
Constante ajoutée.

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,138 +1,169 @@
       getNextPageParam(lastPage) {
-        return lastPage.nextCursor;
+        if (lastPage.next == null) return null;
+        const nextCursor = parseInt(getTokenFromUrl(lastPage.next, 6), 10);
```
6 , 10 -> pas de magic number, créer une constante

Matthieu-EY replied:
Une fonction plus propre, basée sur URL.parse, remplace getTokenFromUrl et enlève au passage l'usage de magic number.
Pour parseInt et 10, ceci est modifié dans la branche feat/pokemon-team, aussi ce n'est pas un problème.
Sur le code final, il n'y aura plus aucun magic number.

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,138 +1,169 @@
+  const [searchedName, setSearchedName] = useState('');
+  const [searchedType, setSearchedType] = useState('All');
+
+  const pokemonsQuery = trpc.poke.list.useInfiniteQuery(
```
On pourrait déstructurer cette objet. Si tu veux, renommer certains propriétés que tu juges pas assez explicite. Mais cela permettrait d'éviter la redondance: pokemonsQuery.XXX

Matthieu-EY replied:
C'est peut-être une affaire de préférence personnelle, mais je trouve que `pokemonsQuery.isLoading` est plus clair que `isPokemonLoading` ou tout autre nom de ce type.
Je préfère donc garder la redondance, à moins d'être obligé du contraire.

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,138 +1,169 @@
-      // refetches posts after a post is added
-      await utils.post.list.invalidate();
-      setFormData({ title: '', text: '' });
+  const lastElementRef = useCallback(
```
Créer un custom hook générique qui, à partir d'une ref, observe l'intersection.
Ici, ça crée de la complexité dans ce composant, alors que cela pourrait être abstrait en dehors de celui-ci.

Matthieu-EY replied:
Custom hook générique créé.

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,138 +1,169 @@
+  const [searchedName, setSearchedName] = useState('');
+  const [searchedType, setSearchedType] = useState('All');
+
+  const pokemonsQuery = trpc.poke.list.useInfiniteQuery(
```
Ta solution de fetching présente certains soucis, saurais-tu me les trouver ? 
Avec l'ajout des fonctionnalités de filtrage, elle pourrait être améliorer. Quelles seraient les solutions ?

Matthieu-EY replied:
Je ne vois pas trop le problème avec le *fetching.*
Ok pour le filtrage, point que j'ai abordé dans d'autres conversations.
Mais ici, je ne vois pas vraiment ce qui est recommandé ou ce qui pourrait être amélioré, @RubenDeAndrade

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,138 +1,169 @@
-                !postsQuery.hasNextPage || postsQuery.isFetchingNextPage
-              }
+        <div className="mt-4 grid justify-center items-center content-center justify-items-center grid-flow-row grid-cols-[repeat(auto-fit,150px)] auto-rows-auto gap-4">
+          {filteredPokemons.map(({ data: pokemon }, index) => (
```
Ici, j'aimerai une liste d'éléments de type Pokemon. Pas d'undefined.

Matthieu-EY replied:
Un tri (filter) a été utilisé pour bien s'assurer de ne pas avoir d'undefined. De plus, un simple mapping a été fait de bien avoir des éléments de type Pokemon.

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,138 +1,169 @@
+              !pokemonsQuery.hasNextPage || pokemonsQuery.isFetchingNextPage
+            }
+          >
+            {pokemonsQuery.isFetchingNextPage
```
J'ai l'impression que ta solution d'infinite fetching, fetch dans tous les cas. Que se passe-t-il si ton filtrage te renvoie seulement 10 éléments. Doit-on afficher le bouton load more ?

Matthieu-EY replied:
Oui, le bouton load more est affiché.
Dès qu'il est visible à l'utilisateur, l'observateur lance une requête pour créer un effet de liste infinie.
Cependant, la liste infinie ne fetch pas continuellement dès le 1er rendu. Il est nécessaire que l'utilisateur scrolle pour que la requête se lance.

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,138 +1,169 @@
+          pokemonsQuery.hasNextPage &&
+          !pokemonsQuery.isFetching
+        ) {
+          pokemonsQuery.fetchNextPage();
```
Ici, il y a une promesse.
J'ai l'impression que tu n'as pas l'extension Eslint d'activer sur ton VSCode.

Matthieu-EY replied:
Résolu.

RubenDeAndrade commented on src/libs/poke/index.ts:
```diff
@@ -57,4 +57,4 @@ export async function fetchPOKE<T>({
  */
 export default {
   fetch: fetchPOKE,
```
since FetchPOKE is a generic function for your fetching solution, you could extract pokemon logic on a new abstract function. Your current function could be the blueprint of your pokeapi fetching logic.
also, we use camelCase for function naming, so here: fetchPoke

Matthieu-EY replied:
Okay for camelCase.
I do not understand the first part of the comment though.
Here is a sample use case of the fetchPoke (now renamed) function:
```js
fetchPoke<Pokemon>({
    path: `/api/v2/pokemon/${id}`,
  }),
```
Since this is very simple, how could it be improved ?

RubenDeAndrade replied:
In the ts-doc, you said it's a generic function.  A generic fetch function should work on any API. Here, it only works for PokeAPI, so its more a reusable function.
To make it generic, you could do:
```tsx
const fetcher = (api_base_url: string) => (path, ...) => {}
const fetchPoke = fetcher(POKE_API_URL)

fetchPoke({
  path: `/api/v2/pokemon/${id}`
})
```

Matthieu-EY replied:
A generic `fetchApi` function has been implemented.
The fetchPoke implementation now depends on fetchApi, as seen in the code above.
As an aside, this change now allows the fetchPoke function to be called in a more simplified manner, like so:
```js
fetchPoke(`/api/v2/pokemon/${id}`)
```
Much simpler now.

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,138 +1,169 @@
+  const scrollObserver = useRef<IntersectionObserver>(null);
+
+  // name inside "search by name" input
+  const [searchedName, setSearchedName] = useState('');
```
Comment pourrait-on améliorer le filtrage par nom ?

Matthieu-EY committed:
fix: retours PR feat/pokemon-list

Matthieu-EY committed:
fix: eslint warnings

Matthieu-EY committed:
lint: linting & recommandations implemented

Matthieu-EY committed:
Merge branch 'main' into feat/pokemon-list

Matthieu-EY committed:
chore: renamed fetchPOKE to fetchPoke

Matthieu-EY committed:
feat: serveur components pour page detail pokemon

- Des composants client ont été créés pour ce qui ne peut pas être côté serveur (typiquement, useState et useRef).

Matthieu-EY committed:
fix: fonction getTokenFromUrl remplacée + little fixes

Matthieu-EY committed:
feat: chargement instantané des 50 1ers Pokémons

Cette méthode peut être clairement améliorée, via initialData passé au useInfiniteQuery et potentiellement aussi au trpc.useQueries

Matthieu-EY committed:
feat: type api request & router

Matthieu-EY committed:
feat: pokemon list type

- Pas de requêtes en boucle en cas de tri par type: tous les Pokémons sont chargés d'un coup, via l'API Type de PokeAPI.

Matthieu-EY committed:
feat: generic function fetchApi

- L'implémentation de fetchpoke dépend maintenant de la fonction fetchApi.
- Ceci permet au passage de simplifier la manière dont fetchpoke est appelée, ce qui simplifie la lecture du code.

Matthieu-EY committed:
feat: prefetch des 50 1ers: initialData

- Amélioration significative du mécanisme de prefetch des 50 1ers Pokémons afin d'améliorer le temps de chargement de la page d'accueil.
- InitialData a permit d'être plus appliqué et le code est plus logique

Co-Authored-By: Ruben De Andrade <28705409+RubenDeAndrade@users.noreply.github.com>

RubenDeAndrade commented on src/app/components/Pokemon/PokemonDisplay.tsx:
```diff
@@ -0,0 +1,78 @@
+                type="button"
+                className="flex p-2 gap-x-4 bg-stone-400 rounded-full items-center justify-center"
+              >
+                <img src="/male.svg" className="max-h-[30px]" alt="Male" />
```
Should do:
remplacer tous les img tag par Next Image component
https://nextjs.org/docs/app/api-reference/components/image

Matthieu-EY replied:
Fait.

RubenDeAndrade commented on src/app/components/Pokemon/PokemonInfo.tsx:
```diff
@@ -0,0 +1,17 @@
+interface PokemonInfoProps {
+  name: string;
+  children: React.ReactNode;
```
Should do:
Je t'encourage à ne plus définir toi même le type de `children` mais d'utiliser plutôt `React.PropsWithChildren` type.
ex: 
```jsx
export const Component = (props: PropsWithChildren<ComponentProps>) => 
```

Matthieu-EY replied:
Fait.
Même si je trouve personnellement que déclarer children explicitement est plus clair pour moi, dans ce cas de figure, les 2 sont possibles et donc PropsWithChildren a été choisi.

RubenDeAndrade commented on src/app/components/PokemonList/PokemonListPage.tsx:
```diff
@@ -0,0 +1,113 @@
+      limit: FETCH_LIMIT_POKEMONS,
+    },
+    {
+      getNextPageParam(lastPage) {
```
`initialData` à définir

Matthieu-EY replied:
implémenté, commit était sur feat/pokemon-team.

RubenDeAndrade commented on src/app/components/PokemonList/PokemonListPage.tsx:
```diff
@@ -0,0 +1,113 @@
+      )
+      .flat() ?? [];
+  const pokemonsTrpc = trpc.useQueries(
+    (t) => pokeIds.map((id) => t.poke.getPokemonById({ id })),
```
initialData à définir

Matthieu-EY replied:
C'est fait.
Le commit était sur la branche feat/pokemon-team. Je l'ai ajouté sur cette branche ci.

RubenDeAndrade commented on src/app/page.tsx:
```diff
@@ -1,140 +1,17 @@
-      },
-    },
+export default async function HomePage() {
+  const initialPokemons = await Promise.all(
```
1. fetch la liste des pokemonIds
2. fetch la liste des details des pokemons

Matthieu-EY replied:
Fait, commit était sur feat/pokemon-team

RubenDeAndrade commented on src/app/hooks/useScrollRef.ts:
```diff
@@ -0,0 +1,28 @@
+      if (scrollObserver.current) scrollObserver.current.disconnect();
+
+      scrollObserver.current = new IntersectionObserver((entries) => {
+        // IIFE defined to avoid the constraints of the callback function, which cannot be async
```
tu peux enlever ce commentaire je pense

Matthieu-EY replied:
Done.

RubenDeAndrade commented on src/app/components/PokemonList/PokemonSearch.tsx:
```diff
@@ -0,0 +1,64 @@
+        <p>Search by type</p>
+        <select
+          defaultValue={searchedType ?? "All"}
+          onChange={(e) => setSearchedType(e.target.value as Type | 'All')}
```
might do:
possiblement créer un type guard isSearchedType au lieu de cast

Matthieu-EY replied:
Remplacé par:
```js
onChange={(e) => setSearchedType(isSearchedType(e.target.value) ? e.target.value : "All")}
```
Qui s'assure que la value est bien un type.

RubenDeAndrade commented on src/app/components/PokemonList/PokemonListType.tsx:
```diff
@@ -0,0 +1,54 @@
+
+        <div className="mt-4 grid justify-center items-center content-center justify-items-center grid-flow-row grid-cols-[repeat(auto-fit,150px)] auto-rows-auto gap-4">
+          {filteredPokemons.map((pokemon, index) => (
+            <PokemonCard key={index + 1} pokemon={pokemon} />
```
Should do:
Utiliser l'id du pokemon plutôt

Matthieu-EY replied:
Fait.

RubenDeAndrade commented on src/app/components/PokemonList/PokemonListPage.tsx:
```diff
@@ -0,0 +1,113 @@
+    (t) => pokeIds.map((id) => t.poke.getPokemonById({ id })),
+  );
+
+  const queryNotFinished = pokemonsQuery?.data?.pages?.[0]?.results == undefined;
+  const PokemonDetailsNotAllFetchedForFirstRender = pokemonsQuery?.data?.pages.length === 1 && pokemonsTrpc.some((pokemonTrpc) => pokemonTrpc.status === "pending");
+  // if first render, load initialPokemons
+  // When useInfiniteQuery catches up, replace the initialPokemons with the new ones.
+  const pokemons = queryNotFinished || PokemonDetailsNotAllFetchedForFirstRender ? initialPokemons : 
+    pokemonsTrpc.map((poke) => poke.data).filter((poke) => poke != null);
```
pas nécessaire

Matthieu-EY replied:
retiré, commit était sur feat/pokemon-team

RubenDeAndrade commented on src/app/components/PokemonList/PokemonListPage.tsx:
```diff
@@ -0,0 +1,113 @@
+
+        <div className="mt-4 grid justify-center items-center content-center justify-items-center grid-flow-row grid-cols-[repeat(auto-fit,150px)] auto-rows-auto gap-4">
+          {filteredPokemons.map((pokemon, index) => (
+            <PokemonCard key={index + 1} pokemon={pokemon} />
```
Should do:
Utiliser l'id du pokemon plutôt

Matthieu-EY replied:
Done.

RubenDeAndrade commented on src/app/components/PokemonList/PokemonListPage.tsx:
```diff
@@ -0,0 +1,113 @@
+  );
+
+  if (searchedType !== "All") {
+    return (<PokemonListType searchedName={searchedName} setSearchedName={setSearchedName} searchedType={searchedType} setSearchedType={setSearchedType} />)
```
Dans PokemonListType, tu fetch des pokemons selon le type.
De fait, les fetch exécutés ici sont obsolètes.
Il y a nécessité de restructurer tes composants pour seulement fetch lorsque c'est nécessaire

Matthieu-EY replied:
Maintenant résolu.

RubenDeAndrade commented on src/utils/ratelimit.ts:
```diff
@@ -0,0 +1,22 @@
+  let inThrottle: boolean;
+  return (...args: unknown[]) => {
+    if (inThrottle) return;
+    console.log(new Date(Date.now()).toISOString())
```
enlever les console.log

Matthieu-EY replied:
Fait.

RubenDeAndrade commented on src/utils/ratelimit.ts:
```diff
@@ -0,0 +1,22 @@
+/* eslint-disable @typescript-eslint/no-unsafe-call */
```
enlever ces commentaires qui désactivent les règles ESLint
Il y a quelques éléments à améliorer dans ton code pour que cela fonctionne.

1. Concernant le type Function, ESLint te recommande:
```
The `Function` type accepts any function-like value.
Prefer explicitly defining any function parameters and return type.eslint[@typescript-eslint/no-unsafe-function-type](https://typescript-eslint.io/rules/no-unsafe-function-type)
```
2. Utiliser les Generics de typescript pour les arguments de tes callbacks

Matthieu-EY replied:
Commentaires supprimés et générics utilisés.

Matthieu-EY committed:
fix: retours PR, fixes, types, lint

Matthieu-EY committed:
fix: retours PR 2, fixes, types, lint

Matthieu-EY committed:
fix: retours PR 3, fixes, lint

Matthieu-EY commented:
Tout semble ok.
Il serait intéressant de merger cette PR et d'en ouvrir une autre pour feat/pokemon-team.
